sdk/generic-sdk-core/generic-core/src/main/java/com/generic/core/http/client/DefaultHttpClient.java

```diff
@@ -156,37 +164,41 @@ private HttpURLConnection connect(HttpRequest httpRequest) {
      * @param connection The HttpURLConnection that is being sent to
      */
     private void sendBody(HttpRequest httpRequest, Object progressReporter, HttpURLConnection connection) {
-        BinaryData binaryDataBody = httpRequest.getBody();
+        BinaryData body = httpRequest.getBody();
 
-        if (binaryDataBody != null) {
-            switch (httpRequest.getHttpMethod()) {
-                case GET:
-                case HEAD: {
-                    return;
-                }
-                case OPTIONS:
-                case TRACE:
-                case CONNECT:
-                case POST:
-                case PUT:
-                case DELETE: {
-                    connection.setDoOutput(true);
-
-                    byte[] bytes = binaryDataBody.toBytes();
-
-                    try (DataOutputStream os = new DataOutputStream(
-                        new BufferedOutputStream(connection.getOutputStream()))) {
-                        os.write(bytes);
-                        os.flush();
-                    } catch (IOException e) {
-                        throw LOGGER.logThrowableAsError(new RuntimeException(e));
-                    }
+        if (body != null) {
+            HttpMethod method = httpRequest.getHttpMethod();
+            if (METHODS_WITHOUT_BODY.contains(method)) {
+                return;
+            }
 
-                    return;
+            if (!METHODS_WITH_BODY.contains(method)) {
+                throw LOGGER.logThrowableAsError(new IllegalStateException("Unknown HTTP Method: " + method));
+            }
+
+            connection.setDoOutput(true);
+
+            if (body instanceof ByteArrayBinaryData || body instanceof ByteBufferBinaryData
+                || body instanceof SerializableBinaryData || body instanceof StringBinaryData) {
```

Comment:
[to consider] what do you think about having a `BinaryData.isBuffered` for this check? Otherwise we tend to forget to update all the places where it's done

Reply:
or actually, something like `BinaryData.copyTo(OutputStream)` can do the optimization in the implementation

Reply:
Let's look at adding both. Possibly add on `BinaryData.copyTo(Channel)` as well. Though that API name may be a bit confusing as does `copyTo` buffer if the `BinaryData` isn't buffered?